core/lib/types/src/storage/writes/compression.rs

```diff
@@ -0,0 +1,434 @@
+use super::*;
+
+// Trait used to define functionality for different compression modes. Defines functions for
+// output size, what type of operation was performed, and value/extended compression.
+trait CompressionMode: 'static {
+    /// Id of the operation being performed.
+    fn operation_id(&self) -> usize;
+    /// Number of bytes the compressed value requires.
+    fn output_size(&self) -> Option<usize>;
+    /// Compress the value.
+    fn compress_value_only(&self) -> Option<Vec<u8>>;
+    /// Concatenation of the metadata byte (5 bits for len and 3 bits for operation type) and the compressed value.
+    fn compress_extended(&self) -> Option<Vec<u8>>;
```

Comment:
not that important but perhaps would be cleaner if we added the metadata byte on the caller side. 

That is, we'd have a `find_best_compressor` method that would return an instance of `CompressionMode`, and the caller would invoke `.operation_id()` on it to detemine the metadata byte. Also, we wouldn't need to duplicate the somewhat convoluted `((self.output_size().unwrap() << 3)` expression

Reply:
I see you resolved it - did you consider doing this? I don't insist, just wanted to make sure you didn't overloop this comment

Reply:
yup, I kept the original code but within each I abstracted out metadata byte creation to a function. I agree that overall it looks cleaner and provides 1 place to change rather than `N`